### PR TITLE
interlis import<->export

### PR DESCRIPTION
**Pull request by olivierdalang**
This work in progress PR shows the current implementation for QGEP/QWAT<->interlis import/export.

It consists of both a command line utility and a GUI integrated in QGEP plugin.

For now, just the case for QGEP is done (QWAT is only a stub).

The GUI is generic, in that it works in the same way whether we are importing QGEP or QWAT data, and whether it is just VSA_KEK data (from Wincan) or a broader SIA405 import.

For now, it just displays some specific options for :
- examinations (allowing to attach the examination to one or more wastewater structures)
- data_media (allowing to rename the path of the media)

The idea is to extend this with more options where adaptation on import is needed.

## Command line usage

```
# import interlis into QGEP
python -m qgepplugin.interlis io --import_xtf qgepplugin/interlis/data/test_data/case_a_import_from_wincan.xtf qgep --recreate_schema

# export QGEP to interlis
python -m qgepplugin.interlis io --export_xtf output.xtf qgep --recreate_schema
```

## Plugin usage

Note : requires ili2pg 4.4.6 or newer. This will be installed automatically through dependency to modelbaker, but for now, must manually replace the ili2pg binary from modelbaker's installation directory with 4.4.6 release (modelbaker installs 4.4.4).

Load the demo QGIS project. Click on "import", and try with `qgepplugin/interlis/data/test_data/case_a_import_from_wincan.xtf`.

Export currently hangs due to some deadlock (not sure why ?! it seems to work when called from command line, and it uses the same functions).

## Tests
```
python -m unittest interlis.tests
```

## Next steps

Priority
- [x] GUI for examination
   - [x] include list of children damage (with a checkbox to toggle import)
   - [x] global filtering of damage (hard-coded, to disable by default codes BCD)

Next steps (these will be done as subsequent improvements, mainly in the qgepqwat2ili repo)
- [ ] split the QGEP and QWAT related code into corresponding repositories
- [x] fix deadlock on export from with QGIS (why is this happening ?! it works with command line)
- [x] improve performance (probably N+1 queries almost everywhere makes import/export very slow)
- [x] manage errors
- [ ] improve UI
    - [ ] improve specific UI (determine what else exactly we need)
    - [ ] probably don't expose the whole list for items without import options but just total count
    - [ ] better progress dialog (incl. propper logs instead of print)
- [ ] deal with dependencies (auto-install sqlalchemy and geoalchemy2)
- [ ] add delta to datamdels to make constraints deferrable (see `qgepplugin/qgepplugin/interlis/qgep/import_.py`)
- [ ] documentation
- [ ] do the whole work on QWAT too
- [ ] include unit tests in CI
- [ ] cleanup
    - [ ]  remove unrelated files (sql/*, datamodel tool related things)
    - [ ] squash commit history

### CODE REVIEW

File: qgepplugin/interlis/NOTES_INTERLIS_QGEP.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Interlis / QGEP oddities
+
+## QGEP
+
+- `qgep_od.file.fk_data_media` doesn't have a FK constraint to `qgep_od.data_media.obj_id`
+
+## SIA405_Abwasser
+
+- `Deckel.Fabrikat` looks like a typo, shouldn't it be `Deckel.Fabrikant` ? (comment is `Name der Herstellerfirma`)
```
Comment (sjib) [inline qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:9]: @olivierdalang German is the leading language - we have to correct the English translation - can you make an issue?

File: qgepplugin/interlis/NOTES_INTERLIS_QGEP.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+## QGEP-SIA405 differences
+
+- `qgep_od.maintenance_event` has a N-M relation to `qgep_od.wastewater_structure`, making fields like `inspection.inspected_length` or `damage.distance` ambigous. Corresponding SIA405_Abwasser classes have a N-1 relation instead, which looks like a better fit.
+
+- In SIA405, some common fields from the concrete `normschachtschaden` and `kanalschaden` could be moved to the abstract `schaden` class (e.g. `distanz`, `quantifizierung1`, `quantifizierung2`...). This would match QGEP's implementation where it's like this already.
```
Comment (sjib) [inline qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:15]: No, we have to adapt QGEP (again) in the way that it is in VSA-KEK - needs a few updates on the data model - I am working on these patches. Even if they look the same - some attributes are MANDATORY in one of the subclasses - so it needs to stay there. @olivierdalang I reopened an issues for that: https://github.com/QGEP/QGEP/issues/153#issuecomment-819616572

File: qgepplugin/interlis/NOTES_INTERLIS_QGEP.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+- `qgep_od.maintenance_event.active_zone` has no equivalent in the interlis model.
```
Comment (sjib) [inline qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:17]: I don't know where this attribute is coming from - should be an QGEP extension attribute (_active_zone) - check who added this in the history of the datamodel

File: qgepplugin/interlis/NOTES_INTERLIS_QGEP.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+- `reach.elevation_determination` has no equivalent in the interlis model
```
Comment (sjib) [inline qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:19]: Corresponds to Haltung.Hoehenbestimmung

File: qgepplugin/interlis/NOTES_INTERLIS_QGEP.md (hunk 3 above)
Comment (urskaufmann) [reply qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:17]: I think this was an extension for Arbon (ask Koni Fischer). I never used something like this.

File: qgepplugin/interlis/NOTES_INTERLIS_QGEP.md (hunk 1 above)
Comment (sjib) [reply qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:9]: I re-checked the translation: Deckel.Fabrikat is translated as cover.brand in deepl (brand as an alternative valid translation) https://www.deepl.com/translator#de/en/Fabrikat. So I think this is ok like this and no correction is needed. e.g. Nestlé can be a company that produces goods and be also a brand.

File: qgepplugin/interlis/NOTES_INTERLIS_QGEP.md (hunk 4 above)
Comment (olivierdalang) [reply qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:19]: I don't see this field in SIA405_Abwasser_2015_2_d-20180417.ili ?

Comment (sjib) [reply qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:19]: It's in the 3D extension

Comment (sjib) [reply qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:19]: `CLASS Haltung EXTENDS SIA405_ABWASSER_2015.SIA405_Abwasser.Haltung =  
    Hoehenbestimmung: (          !! Definiert die Hoehenbestimmung einer Haltung.
      genau,  !! Mittlerer Fehler: Sigma <= +/- 6 cm, Toleranz 3*Sigma <= +/- 18 cm
      unbekannt,
      ungenau  !! Siehe ungenau
    );
    Verlauf3D: Base.Polyline3D;  !! Anfangs-, Knick- und Endpunkte der Leitung (3D Koordinaten);
END Haltung;`

File: qgepplugin/interlis/NOTES_INTERLIS_QGEP.md (hunk 3 above)
Comment (m-kuhn) [reply qgepplugin/interlis/NOTES_INTERLIS_QGEP.md:17]: Yes, that was for Arbon and is a QGEP extension

**Comment by olivierdalang** [comment]
> That is a good temporary fix, but things like this we should fix in a pre-process step where the user gets information, that such things are needed to be changed. The fix with the dimension should not be done automatically

Just to clarify, this only affects the database used in the automated tests (as we need the process succeed to process to succeed in the first place to detect subsequent failures). By no way we should do this type of automatic clamping/fixing on actual data.

Currently, the user gets the warning as an ilivalidator error (ilivalidator is run by default after export). This indeed means we need to do the export before knowing about issues. If we wanted to get an earlier warning, we could either do that in QGEP (but enforcing hard SIA405-specific constraints in QGEP may be quite harsh on users), or have ili2pg create according constraints in the datamodel (but that would completely prevent exporting an invalid model).

**Comment by sjib** [comment]
> Just to clarify, this only affects the database used in the automated tests (as we need the process succeed to process to succeed in the first place to detect subsequent failures). By no way we should do this type of automatic clamping/fixing on actual data.

> Currently, the user gets the warning as an ilivalidator error (ilivalidator is run by default after export). This indeed means we need to do the export before knowing about issues. If we wanted to get an earlier warning, we could either do that in QGEP (but enforcing hard SIA405-specific constraints in QGEP may be quite harsh on users), or have ili2pg create according constraints in the datamodel (but that would completely prevent exporting an invalid model).

Ok. then lets disucss this issue maybe in a separate meeting tomorrow or next week

**Comment by olivierdalang** [comment]
once this is merge, also merge https://github.com/QGEP/docs/pulls

**Review by 3nids — APPROVED** [review]

**Pull request merged**